pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,169 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"regexp"
+	"text/template"
+
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	listers "github.com/knative/serving/pkg/client/listers/networking/v1alpha1"
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler/namespace/config"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/sets"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubelisters "k8s.io/client-go/listers/core/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+// Reconciler implements controller.Reconciler for Certificate resources.
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	nsLister            kubelisters.NamespaceLister
+	knCertificateLister listers.CertificateLister
+
+	configStore configStore
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconciler implements controller.Reconciler for Namespace resources.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	original, err := c.nsLister.Get(key)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := wildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s", cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+		return err
+	}
+
+	dnsnames := sets.NewString(domain)
+	existingCert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Get(ns.Name, metav1.GetOptions{})
+	if existingCert != nil {
+		dnsnames.Insert(existingCert.Spec.DNSNames...)
```

Comment:
I guess you add this line because you want to make sure the "old" DNS names can still be supported when DomainTemplate is changed. But I am not sure if it is necessary as when DomainTemplate is changed, those "old" DNS names will not be used by Routes any more.
We just need to make sure the new DNS names will be supported by Knative Certificates.

Reply:
You're right, I misunderstood how it worked.